File: client/payload.cpp
#include <cstring>
#include <vector>
#include "payload.h"

template <typename T>
void serialize_member(std::vector<uint8_t>& result, const T& data) {
    const uint8_t* bytes = reinterpret_cast<const uint8_t*>(&data);
    result.insert(result.end(), bytes, bytes + sizeof(T));
}

ResponseEmptyPayload::ResponseEmptyPayload() : size(0) {}

uint32_t ResponseEmptyPayload::get_size() {
  return size;
}

std::vector<uint8_t> ResponseEmptyPayload::to_bytes() {
  std::vector<uint8_t> bytes;
  return bytes;
}

ResponsePayload::ResponsePayload(const std::string& client_id) {
  strncpy(this->client_id, client_id.c_str(), 16);
}

std::string ResponsePayload::get_client_id() {
  return std::string(this->client_id);
}

std::vector<uint8_t> ResponsePayload::to_bytes() {
  std::vector<uint8_t> bytes;
  serialize_member(bytes, this->client_id);
  return bytes;
}

ResponseKeyPayload::ResponseKeyPayload(const std::string& client_id, const std::string& encrypted_key) 
  : ResponsePayload(client_id) 
  , encrypted_key(encrypted_key) {}

std::string ResponseKeyPayload::get_encrypted_key() {
  return encrypted_key;
}

std::vector<uint8_t> ResponseKeyPayload::to_bytes() {
  std::vector<uint8_t> bytes = ResponsePayload::to_bytes();
  serialize_member(bytes, this->encrypted_key);
  return bytes;
}

ResponseReceiveFilePayload::ResponseReceiveFilePayload(const std::string& client_id, const uint32_t& content_size, const std::string& file_name, const uint32_t& cksum) 
  : ResponsePayload(client_id)
  , content_size(content_size)
  , cksum(cksum) {
  strncpy(this->file_name, file_name.c_str(), 255);
}

uint32_t ResponseReceiveFilePayload::get_content_size() {
  return content_size;
}

std::string ResponseReceiveFilePayload::get_file_name() {
  return std::string(this->file_name);
}

uint32_t ResponseReceiveFilePayload::get_cksum() {
  return cksum;
}

std::vector<uint8_t> ResponseReceiveFilePayload::to_bytes() {
  std::vector<uint8_t> bytes = ResponsePayload::to_bytes();
  serialize_member(bytes, this->content_size);
  serialize_member(bytes, this->file_name);
  serialize_member(bytes, this->cksum);
  return bytes;
}

RequestUserPayload::RequestUserPayload(const std::string& name) {
  strncpy(this->name, name.c_str(), 255);
}

std::string RequestUserPayload::get_name() {
  return std::string(this->name);
}

std::vector<uint8_t> RequestUserPayload::to_bytes() {
  std::vector<uint8_t> bytes;
  serialize_member(bytes, this->name);
  return bytes;
}

RequestFilePayload::RequestFilePayload(const std::string& file_name) {
  strncpy(this->file_name, file_name.c_str(), 255);
}

std::string RequestFilePayload::get_file_name() {
  return std::string(this->file_name);
}

std::vector<uint8_t> RequestFilePayload::to_bytes() {
  std::vector<uint8_t> bytes;
  serialize_member(bytes, this->file_name);
  return bytes;
}

RequestPublicKeyPayload::RequestPublicKeyPayload(const std::string& name, const std::string& public_key) 
  : RequestUserPayload(name) {
    strncpy(this->public_key, public_key.c_str(), 160);
}

std::string RequestPublicKeyPayload::get_public_key() {
  return std::string(this->public_key);
}

std::vector<uint8_t> RequestPublicKeyPayload::to_bytes() {
  std::vector<uint8_t> bytes = RequestUserPayload::to_bytes();
  serialize_member(bytes, this->public_key);
  return bytes;
}

RequestSentFilePayload::RequestSentFilePayload(const std::string& file_name, const uint32_t& content_size, const std::string& message_content)
  : RequestFilePayload(file_name)
  , content_size(content_size)
  , message_content(message_content) {}

uint32_t RequestSentFilePayload::get_content_size() {
  return this->content_size;
}

std::string RequestSentFilePayload::get_message_content() {
  return this->message_content;
}

std::vector<uint8_t> RequestSentFilePayload::to_bytes() {
  std::vector<uint8_t> bytes = RequestFilePayload::to_bytes();
  serialize_member(bytes, this->content_size);
  serialize_member(bytes, this->message_content);
  return bytes;
}

File: server/payload.py
from abc import ABC

FILE_NAME_SIZE = 255
CLIENT_NAME_SIZE = 255
PUBLIC_KEY_SIZE = 160

class Payload(ABC):
  pass

class RequestUserPayload(Payload):
  def __init__(self, name):
    self.name = name

  def get_name(self):
    return self.name
  
class RequestFilePayload(Payload):
  def __init__(self, file_name):
    self.file_name = file_name

  def get_file_name(self):
    return self.file_name
  
class RequestPublicKeyPayload(RequestUserPayload):
  def __init__(self, name, public_key):
    super().__init__(name)
    self.public_key = public_key

  def get_public_key(self):
    return self.public_key
  
class RequestSentFilePayload(RequestFilePayload):
  def __init__(self, content_size, file_name, message_content):
    super().__init__(file_name)
    self.content_size = content_size
    self.message_content = message_content

  def get_content_size(self):
    return self.content_size
  
  def get_message_content(self):
    return self.message_content
  
class ResponseEmptyPayload(Payload):
  def get_size(self):
    return 0
  
class ResponsePayload(ResponseEmptyPayload):
  def __init__(self, client_id):
    self.client_id = client_id

  def get_client_id(self):
    return self.client_id
  
  def get_size(self):
    # attributes = [attr for attr in dir(self) if not attr.startswith('_') and 
    # not callable(getattr(self, attr))]
    # total_size = sum(getsizeof(getattr(self, attr)) for attr in attributes)
    
    # return total_size
    return 16
  
class ResponseKeyPayload(ResponsePayload):
  def __init__(self, client_id, encrypted_key):
    super().__init__(client_id)
    self.encrypted_key = encrypted_key

  def get_encrypted_key(self):
    return self.encrypted_key
  
  def get_size(self):
    return super().get_size() + len(self.encrypted_key)
  
class ResponseReceiveFilePayload(ResponsePayload):
  def __init__(self, client_id, content_size, file_name, cksum):
    super().__init__(client_id)
    self.content_size = content_size
    self.file_name = file_name
    self.cksum = cksum

  def get_content_size(self):
    return self.content_size
  
  def get_file_name(self):
    return self.file_name
  
  def get_cksum(self):
    return self.cksum
  
  def get_size(self):
    return super().get_size() + 4 + 255 + 4